Exporting requirements raises when the lock file is out of date

Poetry appears to write warnings to stdout when exporting the requirements, causing a parse error.

**Update:** ~~It's likely not Poetry mixing stdout and stderr here. The Nox interface we're using to invoke Poetry does not return these streams separately.~~ It is actually Poetry writing the warning to stdout. And we're already passing `stderr=None` to keep stderr separately.

See this run from #348 where the lock file was not up to date:

```
Run nox --force-color --python=3.9
  nox --force-color --python=3.9
  shell: /usr/bin/bash -e {0}
  env:
    NOXSESSION: pre-commit
    pythonLocation: /opt/hostedtoolcache/Python/3.9.4/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.9.4/x64/lib
nox > Running session pre-commit
nox > Creating virtual environment (virtualenv) using python3.9 in .nox/pre-commit
nox > poetry export --format=requirements.txt --dev --without-hashes
nox > Session pre-commit raised exception RuntimeError('line 1: \'Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.\': Parse error at "\': The lo\'": Expected stringEnd')
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/packaging/requirements.py", line 113, in __init__
    req = REQUIREMENT.parseString(requirement_string)
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/pyparsing.py", line 1955, in parseString
    raise exc
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/pyparsing.py", line 3814, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pyparsing.ParseException: Expected stringEnd, found ':'  (at char 7), (line:1, col:8)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 67, in to_constraint
    requirement = Requirement(requirement_string)
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/packaging/requirements.py", line 115, in __init__
    raise InvalidRequirement(
packaging.requirements.InvalidRequirement: Parse error at "': The lo'": Expected stringEnd

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox/sessions.py", line 549, in execute
    self.func(session)
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox/_decorators.py", line 53, in __call__
    return self.func(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 42, in wrapper
    function(proxy, *_args, **_kwargs)
  File "/home/runner/work/nox-poetry/nox-poetry/noxfile.py", line 81, in precommit
    session.install(
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 276, in install
    return self.poetry.install(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 144, in install
    requirements = self.export_requirements()
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 210, in export_requirements
    constraints = to_constraints(self.poetry.export())
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 89, in to_constraints
    return "\n".join(_to_constraints())
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 85, in _to_constraints
    constraint = to_constraint(requirement, line)
  File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/nox_poetry/sessions.py", line 69, in to_constraint
    raise RuntimeError(f"line {line}: {requirement_string!r}: {error}")
RuntimeError: line 1: 'Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.': Parse error at "': The lo'": Expected stringEnd
nox > Session pre-commit failed.
Error: Process completed with exit code 1.

```

--- a/src/nox_poetry/poetry.py
+++ b/src/nox_poetry/poetry.py
@@ -1,4 +1,5 @@
 """Poetry interface."""
+import sys
 from enum import Enum
 from pathlib import Path
 from typing import Any
@@ -100,6 +101,7 @@ class Poetry:
         def _stripwarnings(lines: Iterable[str]) -> Iterator[str]:
             for line in lines:
                 if line.startswith("Warning:"):
+                    print(line, file=sys.stderr)
                     continue
                 yield line
 

--- a/src/nox_poetry/poetry.py
+++ b/src/nox_poetry/poetry.py
@@ -2,6 +2,8 @@
 from enum import Enum
 from pathlib import Path
 from typing import Any
+from typing import Iterable
+from typing import Iterator
 from typing import List
 from typing import Optional
 
@@ -94,7 +96,14 @@ class Poetry:
             )
 
         assert isinstance(output, str)  # noqa: S101
-        return output
+
+        def _stripwarnings(lines: Iterable[str]) -> Iterator[str]:
+            for line in lines:
+                if line.startswith("Warning:"):
+                    continue
+                yield line
+
+        return "".join(_stripwarnings(output.splitlines(keepends=True)))
 
     def build(self, *, format: str) -> str:
         """Build the package.

--- a/tests/functional/conftest.py
+++ b/tests/functional/conftest.py
@@ -129,10 +129,10 @@ def run_nox_with_noxfile(
     sessions: Iterable[SessionFunction],
     imports: Iterable[ModuleType],
     *nox_args: str,
-) -> None:
+) -> CompletedProcess:
     """Write a noxfile and run Nox in the project."""
     _write_noxfile(project, sessions, imports)
-    _run_nox(project, *nox_args)
+    return _run_nox(project, *nox_args)
 
 
 def list_packages(project: Project, session: SessionFunction) -> List[Package]:

--- a/tests/functional/test_functional.py
+++ b/tests/functional/test_functional.py
@@ -393,3 +393,17 @@ def test_installroot_no_install(project: Project) -> None:
     packages = list_packages(project, test)
 
     assert set(expected) == set(packages)
+
+
+def test_poetry_warnings(datadir: Path) -> None:
+    """It writes warnings from Poetry to the console."""
+    project = Project(datadir / "outdated-lockfile")
+
+    @nox_poetry.session
+    def test(session: nox_poetry.Session) -> None:
+        """Install the local package."""
+        session.install(".")
+
+    process = run_nox_with_noxfile(project, [test], [nox_poetry])
+
+    assert "Warning:" in process.stderr

--- a/tests/functional/test_functional/outdated-lockfile/pyproject.toml
+++ b/tests/functional/test_functional/outdated-lockfile/pyproject.toml
@@ -1,0 +1,12 @@
+[tool.poetry]
+name = "outdated-lockfile"
+version = "0.1.0"
+description = ""
+authors = ["Your Name <your.name@example.com>"]
+
+[tool.poetry.dependencies]
+python = "^3.6.1"
+
+[build-system]
+requires = ["poetry-core>=1.0.0"]
+build-backend = "poetry.core.masonry.api"

--- a/tests/functional/test_functional/outdated-lockfile/src/outdated_lockfile/__init__.py
+++ b/tests/functional/test_functional/outdated-lockfile/src/outdated_lockfile/__init__.py
@@ -1,0 +1,1 @@
+"""Package with an outdated lockfile."""

--- a/tests/unit/test_poetry.py
+++ b/tests/unit/test_poetry.py
@@ -1,5 +1,13 @@
 """Unit tests for the poetry module."""
 from pathlib import Path
+from typing import Any
+from typing import Dict
+
+import nox._options
+import nox.command
+import nox.manifest
+import nox.sessions
+import pytest
 
 from nox_poetry import poetry
 
@@ -16,3 +24,40 @@ name = "África"
 
     config = poetry.Config(path.parent)
     assert config.name == "África"
+
+
+@pytest.fixture
+def session(monkeypatch: pytest.MonkeyPatch) -> nox.Session:
+    """Fixture for a Nox session."""
+    registry: Dict[str, Any] = {}
+    monkeypatch.setattr("nox.registry._REGISTRY", registry)
+
+    @nox.session(venv_backend="none")
+    def test(session: nox.Session) -> None:
+        """Example session."""
+
+    config = nox._options.options.namespace(posargs=[])
+    [runner] = nox.manifest.Manifest(registry, config)
+    runner._create_venv()
+
+    return nox.Session(runner)
+
+
+def test_export_with_warnings(
+    session: nox.Session, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """It removes warnings from the output."""
+    requirements = "first==2.0.2\n"
+    warning = (
+        "Warning: The lock file is not up to date with the latest changes in"
+        " pyproject.toml. You may be getting outdated dependencies. Run update"
+        " to update them.\n"
+    )
+
+    def _run(*args: Any, **kwargs: Any) -> str:
+        return warning + requirements
+
+    monkeypatch.setattr("nox.command.run", _run)
+
+    output = poetry.Poetry(session).export()
+    assert output == requirements

